## Assignments – NumPy package - Random Numbers  
 
### Exercise 1:  
 
Consider a school with10,000 students. Assume a researcher wants to select 100 of them for further study. All their student ids are numbered from 0 to 9999.  
Hint:  
1) Use the following python code to generate a stundentIds list. 
import numpy as np  students_List = range(10000) 
 
2) Use NumPy function, random.choice to randomly select students from the list, students_List. 
 
Choose 100 students randomly without replacement. Use simple random sample method. 
 

In [25]:
import pandas as pd
import numpy as np
from numpy.linalg import inv, qr, det
import statsmodels.api as sm
import os

Exercise 1:
Consider a school with10,000 students. Assume a researcher wants to select 100 of them for further study. All their student ids are numbered from 0 to 9999.

Choose 100 students randomly without replacement. Use simple random sample method.

In [26]:
students_List = range(10000) # generate 10000 student ids 
np.random.seed(101)
rand_hundred = np.random.choice(10000, 100, replace=False) # select random 100 ids without replacement, using simple random method
print("Random 100 student IDs: \n\n", rand_hundred)

Random 100 student IDs: 

 [6676 6421 9834 8492 9982 8844 7247 5862 3662 3457  336 6109 9784 7769
 6971 6001 6995 9758 7115 9925 4421 4150 5342 7122 1185 3814  200 9812
 8444 7925 5310 1010 7778 9579 5905 6847 1549 7335 5966 6270 2351 4864
 1638 9142 4279 3985 8280  800 3106 2630 7699 7347 2161 7242 7761 7874
 9564 8376 3913 3144 2815  166 4995 4819 2054 1009 7393 8080 6871 5646
 2609 7542 1424 6580 8552 3264 4129 7106 5385 3266 2198 3752 2562 9532
 5587 1574 9644 5055 4002 1501 2109 8363 8744  738 5079 1671 2678 3735
 6023 8001]


In [27]:
np.random.seed(101)
rand_hundred1 = np.random.permutation(np.arange(10000))[:100] #another code for the same requirement
print("Random 100 student IDs: \n\n", rand_hundred1)

Random 100 student IDs: 

 [6676 6421 9834 8492 9982 8844 7247 5862 3662 3457  336 6109 9784 7769
 6971 6001 6995 9758 7115 9925 4421 4150 5342 7122 1185 3814  200 9812
 8444 7925 5310 1010 7778 9579 5905 6847 1549 7335 5966 6270 2351 4864
 1638 9142 4279 3985 8280  800 3106 2630 7699 7347 2161 7242 7761 7874
 9564 8376 3913 3144 2815  166 4995 4819 2054 1009 7393 8080 6871 5646
 2609 7542 1424 6580 8552 3264 4129 7106 5385 3266 2198 3752 2562 9532
 5587 1574 9644 5055 4002 1501 2109 8363 8744  738 5079 1671 2678 3735
 6023 8001]


## Exercise 2:

Load Breast cancer dataset provided. The target variable, class is a binary class having values Malignant and Benign There are 569 observations and 30 attributes. Class distribution is 212 Malignant and 357 benign (37% and 63%).

Data set: boston_house_prices.csv

### Split this data set into training and test data set and having the same proportion of classes in the complete data set.

In [28]:
from sklearn import datasets 
from sklearn.model_selection import train_test_split

In [29]:
cancer = datasets.load_breast_cancer() # load the cancer dataset from sklearn datasets
cancer.data

array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]])

In [30]:
cancer.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [31]:
infile = pd.DataFrame(data = cancer['data'] , columns = cancer['feature_names']) # save to dataframe 
infile.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension           ...             worst radius  \
0                 0.07871           ...                    25.38   
1                 0.05667           ...                    24.99   
2                 0.05999           ...                    23.57   
3                 0.09744           ...                    14.91   
4                 0.05883           ...                    22.54   

   worst texture  worst perimeter  worst area  worst smoothness  \
0          17.33           184.60      2019.0            0.1622   
1          23.41           158.80      1956.0            0.1238   
2          25.53           152.50      1709.0            0.1444   
3          26.50            98.87       567.7            0.2098   
4          16.67           152.20      1575.0            0.1374   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   worst fractal dimension  
0                  0.11890  
1                  0.08902  
2                  0.08758  
3                  0.17300  
4                  0.07678  

[5 rows x 30 columns]

In [32]:
# define X and y variables for test split
X = infile # all variables excluding the target variable
y = pd.Series(cancer.target) 

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state= 101, stratify = y) 
# split the original data into train and test subsets using random state and stratify as yes

In [34]:
yval = y.value_counts()
yvalp = y.value_counts(normalize=True)
baseline_class_proportion = pd.concat([yval, yvalp], axis=1)
print("BASELINE proportion of classes: \n", baseline_class_proportion)

BASELINE proportion of classes: 
      0         1
1  357  0.627417
0  212  0.372583


In [35]:
ytrn = y_train.value_counts()
ytrnp = y_train.value_counts(normalize=True)
train_class_proportion = pd.concat([ytrn, ytrnp], axis=1)
print("Proportion of classes in TRAIN dataset: \n", train_class_proportion)

Proportion of classes in TRAIN dataset: 
      0         1
1  250  0.628141
0  148  0.371859


In [36]:
ytst = y_test.value_counts()
ytstp = y_test.value_counts(normalize=True)
train_class_proportion = pd.concat([ytst, ytstp], axis=1)
print("Proportion of classes in TEST dataset: \n", train_class_proportion)

Proportion of classes in TEST dataset: 
      0         1
1  107  0.625731
0   64  0.374269


The above proportions show that the train and test data are having the same proportion of classes as that of the original dataset


## Exercise 3:
Load the pima-indians-diabetes.data.csv provided to you. 

Note: The dataset has no header information.

There are 267 observations belonging to class 1 out of 767 observations. 
Split this data set into training and test data set and having the same proportion of classes in the complete data set. 

In [37]:
os.chdir(r'C:\Users\rr500068\Desktop\Random Numbers\data')
names=['preg','plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
diadata = pd.read_csv('pima-indians-diabetes.data.csv', header=None, names = names)

In [38]:
diadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
preg     768 non-null int64
plas     768 non-null int64
pres     768 non-null int64
skin     768 non-null int64
test     768 non-null int64
mass     768 non-null float64
pedi     768 non-null float64
age      768 non-null int64
class    768 non-null int64
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [39]:
diadata['class'].value_counts()

0    500
1    268
Name: class, dtype: int64

In [40]:
X = diadata.drop('class', axis=1)
y = diadata['class']

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state= 100)
# split the original data into train and test subsets using random state and stratify as yes

In [42]:
yval = y.value_counts()
yvalp = y.value_counts(normalize=True)
baseline_class_proportion = pd.concat([yval, yvalp], axis=1)
print("BASELINE proportion of classes for Diabetes: \n", baseline_class_proportion)

BASELINE proportion of classes for Diabetes: 
    class     class
0    500  0.651042
1    268  0.348958


In [43]:
ytrn = y_train.value_counts()
ytrnp = y_train.value_counts(normalize=True)
train_class_proportion = pd.concat([ytrn, ytrnp], axis=1)
print("Proportion of Diabetes classes in TRAIN dataset: \n", train_class_proportion)

Proportion of Diabetes classes in TRAIN dataset: 
    class     class
0    350  0.651769
1    187  0.348231


In [44]:
ytst = y_test.value_counts()
ytstp = y_test.value_counts(normalize=True)
train_class_proportion = pd.concat([ytst, ytstp], axis=1)
print("Proportion of Diabetes classes in TEST dataset: \n", train_class_proportion)

Proportion of Diabetes classes in TEST dataset: 
    class     class
0    150  0.649351
1     81  0.350649


The above proportions show that the train and test data are having the same proportion of classes as that of the original dataset

## Exercise 4:
ABC company wants to study into the impact of leadership style on employee motivation level and reduction in attrition. ABC company has 240 employees in Operations department who could be potentially interviewed. You identified the sample size as 24 employees.

### Select 10 employees randomly for interviewing.

In [45]:
n = 24
np.random.seed(1234)
k = np.random.randint(1, n, size = 1) 
print('First random number %d'%k)

First random number 16


In [46]:
n = 24
k = np.random.randint(1, n, size = 1) 
np.random.seed(101)

sample_sizes = map(lambda x: x * 24 + k, range(0,10))
random_nos = list(map(int,sample_sizes))
print(random_nos)

[20, 44, 68, 92, 116, 140, 164, 188, 212, 236]


## Exercise 5
ASSUME THERE ARE 55000 CUSTOMERS AND THEIR LIST IS ARRANGED IN THE DECREASING ORDER OF ORDER VOLUMES. 
### You are asked to select 250 customers.

YOU DIVIDE THE POPULATION SIZE BY SAMPLE SIZE = 55000 / 250 = 220.

We have got the population of customers subdivided into 250 customer samples of size 220. So, the sample size, n= 220. So we need to first choose a random number between 1 and 220 (both inclusive) and add the sample size to get other numbers. Now you choose a number randomly between 1 and 220. Assume that number is 48. You would choose 48, 268, 488, and so on.

In [47]:
import numpy as np
n = 220
np.random.seed(101)
print('\nFirst random number is: %d' %k)


First random number is: 20


In [48]:
sample_sizes = map(lambda x: x * 220 + k, range(0,250))
random_nos = list(map(int,sample_sizes))
print('List of random numbers\n')
print(random_nos)
print('\nSample size = ', len(random_nos))

List of random numbers

[20, 240, 460, 680, 900, 1120, 1340, 1560, 1780, 2000, 2220, 2440, 2660, 2880, 3100, 3320, 3540, 3760, 3980, 4200, 4420, 4640, 4860, 5080, 5300, 5520, 5740, 5960, 6180, 6400, 6620, 6840, 7060, 7280, 7500, 7720, 7940, 8160, 8380, 8600, 8820, 9040, 9260, 9480, 9700, 9920, 10140, 10360, 10580, 10800, 11020, 11240, 11460, 11680, 11900, 12120, 12340, 12560, 12780, 13000, 13220, 13440, 13660, 13880, 14100, 14320, 14540, 14760, 14980, 15200, 15420, 15640, 15860, 16080, 16300, 16520, 16740, 16960, 17180, 17400, 17620, 17840, 18060, 18280, 18500, 18720, 18940, 19160, 19380, 19600, 19820, 20040, 20260, 20480, 20700, 20920, 21140, 21360, 21580, 21800, 22020, 22240, 22460, 22680, 22900, 23120, 23340, 23560, 23780, 24000, 24220, 24440, 24660, 24880, 25100, 25320, 25540, 25760, 25980, 26200, 26420, 26640, 26860, 27080, 27300, 27520, 27740, 27960, 28180, 28400, 28620, 28840, 29060, 29280, 29500, 29720, 29940, 30160, 30380, 30600, 30820, 31040, 31260, 31480, 31700, 31920, 32140